### PR TITLE
v2.2.0

### PR DESCRIPTION
**Pull request by benjamc**
# 2.2.0

## Features
- Add example to specify total budget (fidelity units) instead of n_trials for multi-fidelity/Hyperband (#1121)

## Dependencies
- Update numpy NaN (#1122) and restrict numpy version
- Upgrade to ConfigSpace 1.x.x (#1124)

### CODE REVIEW

File: smac/acquisition/maximizer/local_and_random_search.py [hunk 1]
```diff
@@ -69,6 +77,28 @@ def __init__(
             seed=seed,
         )
 
+        if uniform_configspace is not None and prior_sampling_fraction is None:
```
Comment (dengdifan) [inline smac/acquisition/maximizer/local_and_random_search.py:80]: This is replicated with the following lines (starting 102)?

File: smac/acquisition/maximizer/local_and_random_search.py [hunk 2]
```diff
@@ -103,6 +133,8 @@ def __init__(
         self._local_search_iterations = local_search_iterations
         self._prior_sampling_fraction = prior_sampling_fraction
         self._uniform_configspace = uniform_configspace
+        self._prior_sampling_fraction = prior_sampling_fraction
```
Comment (dengdifan) [inline smac/acquisition/maximizer/local_and_random_search.py:136]: Same here, replicated with above lines

File: smac/acquisition/maximizer/local_and_random_search.py [hunk 3]
```diff
@@ -112,6 +144,11 @@ def acquisition_function(self) -> AbstractAcquisitionFunction | None:  # noqa: D
     @acquisition_function.setter
     def acquisition_function(self, acquisition_function: AbstractAcquisitionFunction) -> None:
         self._acquisition_function = acquisition_function
+        if self._uniform_configspace is not None:
```
Comment (dengdifan) [inline smac/acquisition/maximizer/local_and_random_search.py:147]: Same here

File: smac/acquisition/maximizer/local_and_random_search.py [hunk 4]
```diff
@@ -137,6 +174,21 @@ def meta(self) -> dict[str, Any]:  # noqa: D102
                     "local_search": self._local_search.meta,
                 }
             )
+        if self._uniform_configspace is None:
```
Comment (dengdifan) [inline smac/acquisition/maximizer/local_and_random_search.py:177]: Same

File: smac/acquisition/maximizer/local_and_random_search.py [hunk 5]
```diff
@@ -172,6 +223,9 @@ def _maximize(
                 _sorted=True,
             )
 
+        # Choose the best self._local_search_iterations random configs to start the local search, and choose only
```
Comment (dengdifan) [inline smac/acquisition/maximizer/local_and_random_search.py:226]: Same

File: smac/intensifier/hyperband_budget_calculation.ipynb [hunk 6]
```diff
@@ -0,0 +1,73 @@
+{
```
Comment (dengdifan) [inline smac/intensifier/hyperband_budget_calculation.ipynb:1]: Why do we have a jupyter notebook file under the package?

File: smac/acquisition/maximizer/local_and_random_search.py (hunk 5 above)
Comment (benjamc) [reply smac/acquisition/maximizer/local_and_random_search.py:226]: will fix it. this was a merge error

File: smac/intensifier/hyperband_budget_calculation.ipynb (hunk 6 above)
Comment (benjamc) [reply smac/intensifier/hyperband_budget_calculation.ipynb:1]: Is not needed anymore, will remove

File: smac/utils/numpyencoder.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
+from __future__ import annotations
+
+from typing import Any
+
+import json
+
+import numpy as np
+
+
+class NumpyEncoder(json.JSONEncoder):
```
Comment (dengdifan) [inline smac/utils/numpyencoder.py:10]: Do we have a unit test regarding this?

File: smac/intensifier/hyperband_utils.py [hunk 8]
```diff
@@ -0,0 +1,181 @@
+from __future__ import annotations
+
+import numpy as np
+
+from smac.intensifier.successive_halving import SuccessiveHalving
+
+
+def determine_HB(min_budget: float, max_budget: float, eta: int = 3) -> dict:
```
Comment (dengdifan) [inline smac/intensifier/hyperband_utils.py:8]: Unit test missing?

Comment (benjamc) [reply smac/intensifier/hyperband_utils.py:8]: good point, will add

File: smac/utils/numpyencoder.py (hunk 7 above)
Comment (benjamc) [reply smac/utils/numpyencoder.py:10]: no, I can quickly write one

**Review by dengdifan — CHANGES_REQUESTED** [review]
Some lines might need to be deleted and some unit tests are missing. But overall it looks good to me.

File: smac/utils/numpyencoder.py (hunk 7 above)
Comment (benjamc) [reply smac/utils/numpyencoder.py:10]: added

**Review by dengdifan — APPROVED** [review]
Thanks for the PR! LGTM

File: tests/test_intensifier/test_hyperband_utils.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+
+def test_determine_HB():
+    min_budget = 1.0
+    max_budget = 81.0
+    eta = 3
+
+    result = determine_HB(min_budget=min_budget, max_budget=max_budget, eta=eta)
+
+    # Test table 1 from https://arxiv.org/pdf/1603.06560.pdf
```
Comment (helegraf) [inline tests/test_intensifier/test_hyperband_utils.py:17]: It should be noted that the this doesn't actually follow the *table* in the hyperband paper, but instead the *algorithm* (sadly they do not coincide, as noted in #977)

File: smac/intensifier/hyperband_utils.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    total_budget: float, min_budget: float, max_budget: float, eta: int = 3, print_summary: bool = True
+) -> int:
+    """Caculate the number of trials needed for multi-fidelity optimization
+
+    Specify the total budget and find out how many trials that equals.
+
+    Parameters
+    ----------
+    total_budget : float
+        Total budget for the complete optimization in fidelity units
```
Comment (helegraf) [inline smac/intensifier/hyperband_utils.py:166]: could maybe use more explanation on what is meant by a fidelity unit here, maybe a short example in the description

**Review by helegraf — APPROVED** [review]
LGTM except for minor comments

File: tests/test_intensifier/test_hyperband_utils.py (hunk 9 above)
Comment (benjamc) [reply tests/test_intensifier/test_hyperband_utils.py:17]: updated comment

File: smac/intensifier/hyperband_utils.py [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    print("\tnumber of brackets:\t\t\t", hyperband_info["number_of_brackets"])
+    print("\tbudgets per stage:\t\t\t", hyperband_info["budgets_in_stage"])
+    print("\tn configs per stage:\t\t\t", hyperband_info["n_configs_in_stage"])
+    print("-" * (2 * 30 + len("HYPERBAND IN MULTI-FIDELITY") + 2))
+
+
+def get_n_trials_for_hyperband_multifidelity(
+    total_budget: float, min_budget: float, max_budget: float, eta: int = 3, print_summary: bool = True
+) -> int:
+    """Caculate the number of trials needed for multi-fidelity optimization
```
[the remark this answers is not in the document]
Comment (benjamc) [reply smac/intensifier/hyperband_utils.py:159]: fixed

File: smac/intensifier/hyperband_utils.py (hunk 10 above)
Comment (benjamc) [reply smac/intensifier/hyperband_utils.py:166]: added half sentence

**Pull request merged**